### PR TITLE
feat(pipelines) Add manifest only build

### PR DESCRIPTION
**Pull request by bnchrch**
### TL;DR

The following changes were made to the Airbyte CI to add support for manifest-only connectors.

### What changed?

1. **New Connector Language**: Added `MANIFEST_ONLY` as a new connector language in `ConnectorLanguage` enum.
2. **Path Handling**: Introduced `_manifest_only_path` method in `connector_ops/utils.py` to handle paths for manifest-only connectors.
3. **Manifest Path Resolution**: Updated `manifest_path` property in `connector_ops/utils.py` to resolve manifest paths based on file existence.
4. **Build Pipeline**: Created `manifest_only_connectors.py` in `build_image/steps` to define build steps for manifest-only connectors, and added support for it in `pipeline.py` and `__init__.py`.
5. **Testing**: Added `manifest_only_connectors.py` in `test/steps` to define test steps and updated `pipeline.py` and `__init__.py` accordingly.
6. **Tests**: Introduced `test_manifest_only_connectors.py` in `tests/test_build_image` for testing the new functionality.
7. **Version Bump**: Updated the versions of `connector_ops` and `pipelines` packages.

### How to test?

Run CI pipeline to ensure all tests pass.

### CODE REVIEW

**Comment by bnchrch** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39906?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#39907** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39907?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#39906** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39906?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#39905** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39905?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bnchrch and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-ci/connectors/connector_ops/connector_ops/utils.py [hunk 1]
```diff
@@ -261,6 +261,7 @@ class ConnectorLanguage(str, Enum):
     PYTHON = "python"
     JAVA = "java"
     LOW_CODE = "low-code"
+    MANIFEST_ONLY = "manifest-only"
```
Comment (alafanechere) [inline airbyte-ci/connectors/connector_ops/connector_ops/utils.py:264]: We've a lot of `if connector.language in [ConnectorLanguage.PYTHON, ConnectorLanguage.LOW_CODE]` in `airbyte-ci`, I'm afraid this change might silently filter out the `MANIFEST_ONLY` connectors from some `airbyte-ci` commands. 

The [`up-to-date` worfklow](https://github.com/airbytehq/airbyte/blob/master/.github/workflows/connectors_up_to_date.yml#L14) also uses a `--language` filter, can you please update it to run on this new connector language?

File: airbyte-ci/connectors/pipelines/README.md [hunk 2]
```diff
@@ -761,6 +761,7 @@ E.G.: running Poe tasks on the modified internal packages of the current branch:
 
 | Version | PR                                                         | Description                                                                                                                  |
 | ------- | ---------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------- |
+| 4.21.0  | [#TODO](https://github.com/airbytehq/airbyte/pull/TODO)    | Add manifest only build pipeline                                                          |
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/README.md:764]:
```suggestion
| 4.21.0  | [#39906](https://github.com/airbytehq/airbyte/pull/39906)    | Add manifest only build pipeline                                                          |
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/manifest_only_connectors.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    async def _build_from_base_image(self, platform: Platform) -> Container:
+        """Build the connector container using the base image defined in the metadata, in the connectorBuildOptions.baseImage field.
+
+        Returns:
+            Container: The connector container built from the base image.
+        """
+        self.logger.info(f"Building connector from base image in metadata for {platform}")
+        base = self._get_base_container(platform)
+
+        customized_base = await build_customization.pre_install_hooks(self.context.connector, base, self.logger)
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/manifest_only_connectors.py:47]: I think there's duplication between `_build_from_base_image` in this class vs the python one. Can we keep a common code path for setting entrypoints, labels, calling pre/post hooks etc?

Your could might "just be" considered a  pre-install hook?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/manifest_only_connectors.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+from pipelines.airbyte_ci.connectors.build_image.steps.manifest_only_connectors import BuildConnectorImages
+from pipelines.airbyte_ci.connectors.consts import CONNECTOR_TEST_STEP_ID
+from pipelines.airbyte_ci.connectors.test.context import ConnectorTestContext
+from pipelines.airbyte_ci.connectors.test.steps.common import AcceptanceTests, IncrementalAcceptanceTests, LiveTests
+from pipelines.consts import LOCAL_BUILD_PLATFORM
+from pipelines.helpers.execution.run_steps import STEP_TREE, StepToRun
+
+
+def get_test_steps(context: ConnectorTestContext) -> STEP_TREE:
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/manifest_only_connectors.py:15]: 💅 I'm wondering if we could find a way to have a common tree for low_code/manifest/python connectors and just add unit tests / integration tests for low_code and python connectors.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/manifest_only_connectors.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        entrypoint = build_customization.get_entrypoint(self.context.connector)
+
+        manifest_file_name = "manifest.yaml"
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/manifest_only_connectors.py:50]: 💅 This can be stored in the `const` module

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py [hunk 6]
```diff
@@ -100,6 +100,7 @@ async def _run(self, built_containers_per_platform: Dict[Platform, Container]) -
         ], "This step can only run for Python connectors."
         built_container = built_containers_per_platform[LOCAL_BUILD_PLATFORM]
         pip_freeze_output = await built_container.with_exec(["pip", "freeze"], skip_entrypoint=True).stdout()
+        # TODO CDK version for manifest only connectors
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py:103]: Why does it have to change? In any case we might reconsider the source of truth here: the [dependency list from the connector insights](https://storage.googleapis.com/prod-airbyte-cloud-connector-metadata-service/connector_insights/source-recurly/latest/insights.json) is more exhaustive as it's based on the SBOM.

File: airbyte-ci/connectors/pipelines/poetry.lock
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/poetry.lock:1]: not sure why an update of the poetry lock was required.

**Review by alafanechere — APPROVED** [review]
Approving as there's nothing utterly blocking on my end but please:
* Consider sharing as much logic as possible between the "python" connector types. It'll simplify refactor if we have to modify the testing flow or the build flow.
* Can you update all logics in `airbyte-ci`, `qa_checks`, `connectors_insights` which are based on the `ConnectorLanguage`? So far I've mainly used `if connector.language in [ConnectorLanguage.PYTHON, ConnectorLanguage.LOW_CODE]` to filter "python" connectors.

File: airbyte-ci/connectors/connector_ops/connector_ops/utils.py (hunk 1 above)
Comment (bnchrch) [reply airbyte-ci/connectors/connector_ops/connector_ops/utils.py:264]: Can certainly do.

@alafanechere Just to confirm the up-to-date command will bump the base image?

(since these manifest only connectors wont have a pyproject.toml to update)

Comment (alafanechere) [reply airbyte-ci/connectors/connector_ops/connector_ops/utils.py:264]: Yes it will bump the base image

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/manifest_only_connectors.py (hunk 3 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/manifest_only_connectors.py:47]: @alafanechere I was thinking on this. 

Im not sure this is the place to DRY up the code **completely** and have a common build for python and manifest only connectors.

The reasons for that are
1. The similarities between the manifest and python `_build_from_base_image` are that they use pre and post commit hooks. After that the python build does a lot more inbetween with two stage build
2. We are likely going to deviate more from python as we look at how we add support for custom components, potentially in other languages.

I'm almost more inclined to 
1. Remove the hooks for manifest only
2. Add a helper function for adding the docker labels that the two builds can share

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/manifest_only_connectors.py (hunk 4 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/manifest_only_connectors.py:15]: yes, but these changes are in the rather distant future, aren't they? I imagine we'll be reworking the step tree long before then. Anyway, it's not blocking for me, so I'll let you do what you think is wise :)

Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/manifest_only_connectors.py:15]: Theyre not too far away!

Also the way were likely to do "upload for pyairbyte" changes.

So the types of tests we run are even different

https://airbytehq-team.slack.com/archives/C02TL38U5L7/p1719448571131109

Thanks for the push here btw.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/common.py [hunk 7]
```diff
@@ -35,9 +43,10 @@ async def _run(self, *args: Any) -> StepResult:
         build_results_per_platform = {}
         for platform in self.build_platforms:
             try:
-                connector = await self._build_connector(platform, *args)
+                connector_container = await self._build_connector(platform, *args)
+                connector_container = apply_airbyte_docker_labels(connector_container, self.context.connector)
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/common.py:47]: @alafanechere I had an idea to move the application of the docker labels to a generic place for all connectors.

WDYT?
